src/osvr/USBSerial/USBSerialDevInfo_MacOSX.h

```diff
@@ -51,51 +54,87 @@ namespace usbserial {
+         * Helper function that returns an IOKit iterator for all serial ports on the system.
+         * @param matchingServices pointer which will be set to the value of the iterator
+         * @return 0 on success, -1 on failure
          */
-        boost::optional<USBSerialDevice>
-        make_USBSerialDevice(const std::string &device) {
-            //FIXME: STUB
-            return boost::none;
+        int findSerialPorts(io_iterator_t *matchingServices) {
+
+            kern_return_t kernResult;
+            CFMutableDictionaryRef classesToMatch;
+            // Query IOKit for services matching kIOSerialBSDServiceValue
+            classesToMatch = IOServiceMatching(kIOSerialBSDServiceValue);
+            if (classesToMatch == NULL) {
+                return -1;
+            }
+            else {
+                // Query only for serial ports
+                CFDictionarySetValue(classesToMatch,
+                                     CFSTR(kIOSerialBSDTypeKey),
+                                     CFSTR(kIOSerialBSDAllTypes));
+            }
+            // Run the query
+            kernResult = IOServiceGetMatchingServices(kIOMasterPortDefault, classesToMatch, matchingServices);
+            if(KERN_SUCCESS !=kernResult) {
+                return -1;
+            }
+            return 0;
         }
-
-        /**
-         * Predicate that determines if the given vendor and product IDs match
-         * that of the provided USB serial device. If the @param vendorID or
-         * @param productID is @c boost::none, that parameter will match and ID.
-         *
-         * @param device the USB serial device to compare against
-         * @param vendorID optional vendor ID
-         * @param productID optional product ID
-         *
-         * @return true if the vendor and product IDs match, false otherwise
-         */
-        bool matches_ids(const USBSerialDevice &device,
-                         const boost::optional<uint16_t> &vendorID,
-                         const boost::optional<uint16_t> &productID) {
-            const bool vendor_matches =
-                (!vendorID || *vendorID == device.getVID());
-            const bool product_matches =
-                (!productID || *productID == device.getPID());
-
-            return (vendor_matches && product_matches);
-        }
-
-    } // end namespace
+    }
 
     std::vector<USBSerialDevice>
     getSerialDeviceList(boost::optional<uint16_t> vendorID,
                         boost::optional<uint16_t> productID) {
 
-        //FIXME: STUB
         std::vector<USBSerialDevice> devices;
+
+        io_iterator_t serialPortIterator;
+        io_object_t serialPortService;
+        // find all serial ports
+        if(findSerialPorts(&serialPortIterator) != 0) {
+            goto bailout;
+        }
+
+        // iterate over serial ports, getting vid and pid
+        while((serialPortService = IOIteratorNext(serialPortIterator)) != 0) {
```

Comment:
We could drop the `!= 0` here and get the same semantics, but I'm not sure what @rpavlik's take on that is. The way it's written now make it clear.